File: data_structures/week_1/starter_code_PA1/check_brackets_in_code/check_brackets.cpp
#include <iostream>
#include <stack>
#include <string>


struct Bracket {
    Bracket(char type, int position):
        type(type),
        position(position)
    {}

    bool Match(char c) {
        if (type == '[' && c == ']')
            return true;
        if (type == '{' && c == '}')
            return true;
        if (type == '(' && c == ')')
            return true;
        return false;
    }

    char type;
    int position;
};

int main() {
    std::string text;
    // getline(std::cin, text);
    text = "[]";

    std::stack <Bracket> opening_brackets_stack;
    bool flag = false;
    for (int position = 0; position < text.length(); ++position) {
        char next = text[position];

        if (next == '(' || next == '[' || next == '{') {
            opening_brackets_stack.push(Bracket(next, position));

        } else if (next == ')' || next == ']' || next == '}') {
            if (opening_brackets_stack.size() > 0) {
                std::cout << position;
                flag = true;
                break;
            }
            Bracket* top = opening_brackets_stack.pop();
            if (top.Match(next) == false) {
                std::cout << opening_brackets_stack.pop().position;
                flag = true;
                break;
            }
        }
    }

    if (opening_brackets_stack.size() > 0) {
        Bracket* top = opening_brackets_stack.pop();
        std::cout << top.position;
        flag = true;
    }

    if (flag == false) {
        std::cout << "Success";
    }
}

File: data_structures/week_1/starter_code_PA1/check_brackets_in_code/check_brackets.py
# python3

import sys


class Bracket:
    def __init__(self, bracket_type, position):
        self.bracket_type = bracket_type
        self.position = position

    def match(self, c):
        if self.bracket_type == '[' and c == ']':
            return True
        if self.bracket_type == '{' and c == '}':
            return True
        if self.bracket_type == '(' and c == ')':
            return True
        return False


def bracket_check(text):
    opening_brackets_stack = []
    for index, next in enumerate(text, start=1):
        if next in ("[", "(", "{"):
            opening_brackets_stack.append(Bracket(next, index))

        elif next in ("]", ")", "}"):
            if not opening_brackets_stack:
                return index

            top = opening_brackets_stack.pop()
            if not top.match(next):
                return index
    if opening_brackets_stack:
        top = opening_brackets_stack.pop()
        return top.position

    return "Success"


if __name__ == "__main__":
    text = sys.stdin.read().strip("\n")
    print(bracket_check(text))
